python/ray/data/_internal/planner/planner.py

```diff
@@ -1,28 +1,98 @@
-from typing import Dict
+from typing import Callable, Dict, List, Tuple, Type, TypeVar
 
 from ray.data._internal.execution.interfaces import PhysicalOperator
-from ray.data._internal.execution.operators.union_operator import UnionOperator
-from ray.data._internal.execution.operators.zip_operator import ZipOperator
 from ray.data._internal.logical.interfaces import (
     LogicalOperator,
     LogicalPlan,
     PhysicalPlan,
 )
-from ray.data._internal.logical.operators.all_to_all_operator import AbstractAllToAll
-from ray.data._internal.logical.operators.from_operators import AbstractFrom
-from ray.data._internal.logical.operators.input_data_operator import InputData
-from ray.data._internal.logical.operators.map_operator import AbstractUDFMap
-from ray.data._internal.logical.operators.n_ary_operator import Union, Zip
-from ray.data._internal.logical.operators.one_to_one_operator import Limit
-from ray.data._internal.logical.operators.read_operator import Read
-from ray.data._internal.logical.operators.write_operator import Write
-from ray.data._internal.planner.plan_all_to_all_op import plan_all_to_all_op
-from ray.data._internal.planner.plan_from_op import plan_from_op
-from ray.data._internal.planner.plan_input_data_op import plan_input_data_op
-from ray.data._internal.planner.plan_limit_op import plan_limit_op
-from ray.data._internal.planner.plan_read_op import plan_read_op
-from ray.data._internal.planner.plan_udf_map_op import plan_udf_map_op
-from ray.data._internal.planner.plan_write_op import plan_write_op
+from ray.util.annotations import DeveloperAPI
+
+LogicalOperatorType = TypeVar("LogicalOperatorType", bound=LogicalOperator)
+
+PlanLogicalOpFn = Callable[
+    [LogicalOperatorType, List[PhysicalOperator]], PhysicalOperator
+]
+
+# A list of registered plan functions for logical operators.
+PLAN_LOGICAL_OP_FNS: List[Tuple[Type[LogicalOperator], PlanLogicalOpFn]] = []
+
+
+@DeveloperAPI
+def register_plan_logical_op_fn(
+    logical_op_type: Type[LogicalOperator],
+    plan_fn: PlanLogicalOpFn,
+):
+    """Register a plan function for a logical operator type."""
+    PLAN_LOGICAL_OP_FNS.append((logical_op_type, plan_fn))
+
+
+def register_plan_logical_op_fns():
```

Comment:
Nit: Maybe something like this to disambiguate from `register_plan_logical_op_fn`? 
```suggestion
def register_default_plan_logical_op_fns():
```

From the name, I'd expect something like this:
```
register_plan_logical_op_fns(plan_fns: List[PlanLogicalOpFn])
```

Reply:
makes sense